// ttn/dist/applications/normalize.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _defineProperty2 = require("babel-runtime/helpers/defineProperty");

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _extends3 = require("babel-runtime/helpers/extends");

var _extends4 = _interopRequireDefault(_extends3);

var _objectWithoutProperties2 = require("babel-runtime/helpers/objectWithoutProperties");

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _sourceMapSupport2 = require("source-map-support");

exports.default = function (device) {
  var _device$lorawanDevice = device.lorawanDevice,
      lorawanDevice = _device$lorawanDevice === undefined ? {} : _device$lorawanDevice,
      _device$attributesMap = device.attributesMap,
      attributesMap = _device$attributesMap === undefined ? [] : _device$attributesMap,
      rest = (0, _objectWithoutProperties3.default)(device, ["lorawanDevice", "attributesMap"]);
  var appEui = lorawanDevice.appEui,
      devEui = lorawanDevice.devEui,
      devAddr = lorawanDevice.devAddr,
      nwkSKey = lorawanDevice.nwkSKey,
      appSKey = lorawanDevice.appSKey,
      appKey = lorawanDevice.appKey,
      lorawan = (0, _objectWithoutProperties3.default)(lorawanDevice, ["appEui", "devEui", "devAddr", "nwkSKey", "appSKey", "appKey"]);


  return (0, _extends4.default)({}, rest, lorawan, {
    appEui: hex(appEui),
    devEui: hex(devEui),
    appKey: hex(appKey),
    appSKey: hex(appSKey),
    nwkSKey: hex(nwkSKey),
    devAddr: hex(devAddr),
    attributes: attributesMap.reduce(function (acc, pair) {
      return (0, _extends4.default)({}, acc, (0, _defineProperty3.default)({}, pair[0], pair[1]));
    }, {})
  });
};

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

(0, _sourceMapSupport2.install)(); // Copyright © 2017 The Things Network
// Use of this source code is governed by the MIT license that can be found in the LICENSE file.

var hex = function hex(base64) {
  if (!base64) {
    return;
  }

  return new Buffer(base64, "base64").toString("hex");
};
