// ttn/dist/handler/handler.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HandlerClient = undefined;

var _regenerator = require("babel-runtime/regenerator");

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require("babel-runtime/helpers/asyncToGenerator");

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _classCallCheck2 = require("babel-runtime/helpers/classCallCheck");

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require("babel-runtime/helpers/createClass");

var _createClass3 = _interopRequireDefault(_createClass2);

var _sourceMapSupport2 = require("source-map-support");

var _discovery = require("../discovery");

var _applications = require("../applications");

var _data = require("../data");

var _utils = require("../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

(0, _sourceMapSupport2.install)(); // Copyright © 2017 The Things Network
// Use of this source code is governed by the MIT license that can be found in the LICENSE file.

// Necessary to make gRPC work
process.env.GRPC_SSL_CIPHER_SUITES = _utils.MODERN_CIPHER_SUITES;

/**
 * `Handler` is a  client for The Things Network handler APIs.
 * It can be used to get data from an application or to manage devices.
 *
 * Example:
 * ```
 * const handler = new Handler("my-app-id", "my-app-access-key")
 * ```
 */

var HandlerClient = exports.HandlerClient = function () {

  /** @private */

  /** @private */
  function HandlerClient(appID, appAccessKey, opts) {
    (0, _classCallCheck3.default)(this, HandlerClient);

    this.appID = appID;
    this.appAccessKey = appAccessKey;
    this.discoveryOptions = opts;
  }

  /**
   * `open` opens the client to the handler.
   */


  /** @private */


  /** @private */


  (0, _createClass3.default)(HandlerClient, [{
    key: "open",
    value: function () {
      var _ref = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee() {
        var discovery;
        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!this.announcement) {
                  _context.next = 2;
                  break;
                }

                return _context.abrupt("return", this);

              case 2:
                discovery = new _discovery.Discovery(this.discoveryOptions);
                _context.next = 5;
                return discovery.getByAppID(this.appID);

              case 5:
                this.announcement = _context.sent;
                return _context.abrupt("return", this);

              case 7:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function open() {
        return _ref.apply(this, arguments);
      }

      return open;
    }()

    /**
     * Open a data client that can be used to receive live application data
     */

  }, {
    key: "data",
    value: function data() {
      if (!this.announcement) {
        throw new Error("No handler configured, call open()!");
      }

      return new _data.DataClient(this.appID, this.appAccessKey, this.announcement.mqttAddress);
    }

    /**
     * Open a application manager that can be used to manage the settings and devices of the
     * application.
     */

  }, {
    key: "application",
    value: function application() {
      if (!this.announcement) {
        throw new Error("No handler configured, call discover() or configure() first!");
      }

      return new _applications.ApplicationClient(this.appID, this.appAccessKey, this.announcement.netAddress, this.announcement.certificate);
    }
  }]);
  return HandlerClient;
}();
