// ttn/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.account = exports.data = exports.application = exports.services = exports.Discovery = exports.ApplicationClient = exports.wildcard = exports.DataClient = exports.key = exports.HandlerClient = undefined;

var _regenerator = require("babel-runtime/regenerator");

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require("babel-runtime/helpers/asyncToGenerator");

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _sourceMapSupport2 = require("source-map-support");

var _handler = require("./handler");

Object.defineProperty(exports, "HandlerClient", {
  enumerable: true,
  get: function get() {
    return _handler.HandlerClient;
  }
});

var _key = require("./utils/key");

Object.defineProperty(exports, "key", {
  enumerable: true,
  get: function get() {
    return _key.key;
  }
});

var _data = require("./data");

Object.defineProperty(exports, "DataClient", {
  enumerable: true,
  get: function get() {
    return _data.DataClient;
  }
});
Object.defineProperty(exports, "wildcard", {
  enumerable: true,
  get: function get() {
    return _data.wildcard;
  }
});

var _applications = require("./applications");

Object.defineProperty(exports, "ApplicationClient", {
  enumerable: true,
  get: function get() {
    return _applications.ApplicationClient;
  }
});

var _discovery = require("./discovery");

Object.defineProperty(exports, "Discovery", {
  enumerable: true,
  get: function get() {
    return _discovery.Discovery;
  }
});
Object.defineProperty(exports, "services", {
  enumerable: true,
  get: function get() {
    return _discovery.services;
  }
});

var _account = require("./account");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

(0, _sourceMapSupport2.install)(); // Copyright © 2017 The Things Network
// Use of this source code is governed by the MIT license that can be found in the LICENSE file.

/**
 * `open` creates and opens a HandlerClient for the application with the specified ID.
 *
 * @param appID - The ID of the application you want to manage.
 * @param tokenOrKey - The Access Token or Access Key used to authenticate.
 * @param opts - Optional options to pass to the Discovery client.
 */
var open = function () {
  var _ref = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(appID, accessKerOrToken, opts) {
    var handler;
    return _regenerator2.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            handler = new _handler.HandlerClient(appID, accessKerOrToken, opts);
            _context.next = 3;
            return handler.open();

          case 3:
            return _context.abrupt("return", handler);

          case 4:
          case "end":
            return _context.stop();
        }
      }
    }, _callee, this);
  }));

  return function open(_x, _x2, _x3) {
    return _ref.apply(this, arguments);
  };
}();

/**
 * `application` creates and opens an ApplicationClient for the application with the specified ID.
 *
 * @param appID - The ID of the application you want to manage
 * @param accessKeyOrToken - The Access Token or Access Key used to authenticate
 */
var application = exports.application = function () {
  var _ref2 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee2(appID, accessKeyOrToken, opts) {
    var handler;
    return _regenerator2.default.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            _context2.next = 2;
            return open(appID, accessKeyOrToken, opts);

          case 2:
            handler = _context2.sent;
            return _context2.abrupt("return", handler.application());

          case 4:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2, this);
  }));

  return function application(_x4, _x5, _x6) {
    return _ref2.apply(this, arguments);
  };
}();

/**
 * `data` creates and opens an DataClient for the application with the specified ID.
 *
 * @param appID - The ID of the application you want to manage
 * @param accessKeyOrToken  - The Access Token or Access Key used to authenticate
 */
var data = exports.data = function () {
  var _ref3 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee3(appID, accessKeyOrToken, opts) {
    var handler;
    return _regenerator2.default.wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            _context3.next = 2;
            return open(appID, accessKeyOrToken, opts);

          case 2:
            handler = _context3.sent;
            return _context3.abrupt("return", handler.data());

          case 4:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee3, this);
  }));

  return function data(_x7, _x8, _x9) {
    return _ref3.apply(this, arguments);
  };
}();

/**
 * `account` creates an AccountClient for the user associated to the specified key or token.
 *
 * @param accessKeyOrToken  - The Access Token or Access Key used to authenticate
 * @param serverAddress  - The URL to the account server to use. Defaults to "https://account.thethingsnetwork.org"
 */
var account = exports.account = function account(accessKeyOrToken, serverAddress) {
  return new _account.AccountClient(accessKeyOrToken, serverAddress);
};
